### PR TITLE
cancellable assertion

### PR DESCRIPTION
**Pull request by mskonovalov**
`get_assertion` is blocking on user input so there need to be the way to cancel the blocking function.

It is useful when you have several MFA devices plugged and want to run `get_assertion` on all of them in parallel in separate go-routines. then you can get the result from the device that was touched and cancel the rest.
It replicates the behavior of `navigator.credemtials.get()` in web browser API. See here https://github.com/Yubico/libfido2/issues/250 for the context

### CODE REVIEW

**Review by gabriel — DISMISSED** [review]
Using `context.Context` might be difficult to get right, if you need help lmk, I can do a pass.

File: fido2.go [hunk 1]
```diff
@@ -717,6 +732,28 @@ func (d *Device) Assertion(
 	return assertion, nil
 }
 
+func (d *Device) CancellableAssertion(
```
[the remark this answers is not in the document]
Comment (gabriel) [reply fido2.go:735]: So this would be `(d *Device) Assertion(ctx context.Context, ...`.

[the remark this answers is not in the document]
Comment (mskonovalov) [reply fido2.go:735]: Sure, I have an implementation for my needs based on the `context`. Was not sure if you want to keep the lib low-level or not.
But as Assertion blocks the execution it implies creating several goroutines.
I'll send a PR 👍

[the remark this answers is not in the document]
Comment (gabriel) [reply fido2.go:735]: Actually, I think maybe we should expose `fido_dev_cancel` and then we don't need a special `CancellableAssertion`?

[the remark this answers is not in the document]
Comment (mskonovalov) [reply fido2.go:735]: I tried that. But as `open` is private and we cannot open device twice (at least I think that was the issue) you need to expose opened Device from the `Assertion` function or we would need to allow people to open the device and then pass it into `Assertion` and `Cancel` functions

[the remark this answers is not in the document]
Comment (gabriel) [reply fido2.go:735]: Maybe we want to keep the current open `fido_dev_t` in the `Device` instance, and then allow `Cancel` on it.

File: fido2.go [hunk 2]
```diff
@@ -675,6 +687,9 @@ func (d *Device) Assertion(
 
 	// Get assertion
 	if cErr := C.fido_dev_get_assert(dev, cAssert, cStringOrNil(pin)); cErr != C.FIDO_OK {
+		if cErr == C.FIDO_ERR_KEEPALIVE_CANCEL {
```
Comment (gabriel) [inline fido2.go:690]: Don't need this since errFromCode does it.

**Comment by gabriel** [comment]
OK, sorry for all the back and forth.. I think maybe adding a Cancel method which acts on what is currently open.

I made a PR here with what I had in mind:
https://github.com/keys-pub/go-libfido2/pull/21

Does this work for your scenario?

File: fido2.go (hunk 2 above)
Comment (mskonovalov) [reply fido2.go:690]: If we remove that the error will be wrapped so you need to unwrap on the client side to check if it was interrupted.
What do you think is the best approach here?

**Comment by mskonovalov** [comment]
Thanks, I will test this PR and leave the feedback 👍

File: fido2.go (hunk 2 above)
Comment (gabriel) [reply fido2.go:690]: I think it is ok to expect callers should check errors with errors.Cause(err).

**Pull request closed without merging**